flytekit/core/promise.py

```diff
@@ -1214,15 +1220,24 @@ def create_and_link_node(
     # These will be our core Nodes until we can amend the Promise to use NodeOutputs that reference our Nodes
     upstream_nodes = list(set([n for n in nodes if n.id != _common_constants.GLOBAL_INPUT_NODE_ID]))
 
+    # TODO: Better naming, probably a derivative of the function name.
+    # if not adding to compilation state, we don't need to generate a unique node id
+    node_id = (
+        f"{ctx.compilation_state.prefix}n{len(ctx.compilation_state.nodes)}"
+        if add_node_to_compilation_state and ctx.compilation_state
+        else ""
```

Comment:
can we add an additional optional argument to this function for node_id?  and then let's pass in a constant `"ARRAY_NODE_SUB"` or some other constant string?  at least that way it'll be obvious.

Reply:
sounds good. Updated